02-intermediate-ruby/api-post.md

```diff
@@ -0,0 +1,282 @@
+
+VCR.configure do |config|
+  config.cassette_library_dir = 'specs/cassettes' # folder where casettes will be located
+  config.hook_into :webmock # tie into this other tool called webmock
+  config.default_cassette_options = {
+    record: :new_episodes,    # record new data when we don't have it yet
+    match_requests_on: [:method, :uri, :body] # The http method, URI and body of a request all need to match
+  }
+  # Don't leave our Slack token lying around in a cassette file.
+  config.filter_sensitive_data("SLACK_TOKEN") do
+    ENV["SLACK_TOKEN"]
+  end
+end
+```
+
+This code configures VCR just like we did when we tested `GET` requests.  We have added `config.filter_sensitive_data("SLACK_TOKEN")` so that anytime our VCR sees our slack token it will replace it with the words "SLACK_TOKEN".  **This will prevent the token from appearing in git**.  
+
+We can add our Slack token to a `.env` file in the project's root folder like this:
+
+```
+SLACK_TOKEN: <TOKEN_GOES_HERE>
+```
+
+![.env file picture](images/dotenv.png)
+
+Next we can write our first test using VCR.
+
+```ruby
+# slack_api_wrapper_spac.rb
+require_relative 'spec_helper'
+
+describe SlackApiWrapper do
+  it "can send a valid tweet" do
+    VCR.use_cassette("slack-posts") do
+      response = SlackApiWrapper.send_msg("Hey I can post messages!", "YOUR-CHANNEL-NAME")
+      expect(response).must_equal true
+    end
+  end
+end
+```
+
+This test calls a class-method called `send_msg` and sends it a message and channel name.  When we run our tests we get one test error because the method does not yet exist.  Now we can write `send_msg` to get the test to pass.
+
+With `HTTParty`, you can use the `post` method to send `POST` requests, it takes two arguments.
+
+1. The URL of the post request, like `https://slack.com/api/chat.postMessage`.  
+1. A hash with the following key-value pairs.
+    - `headers` a hash of key-value pairs to put in the request header.
+    - `body` a hash of key-value pairs for the data sent to the API.  
+
+Like the `get` method, `post` returns an HTTP response from the API.  
+
+For example:  
+
+```ruby
+HTTParty.post("http://adapets.org/pets", 
+  headers: { 'Content-Type' => 'application/json' },
+  body: {
+    name: 'Kylo'
+    human: 'Kari',
+    species: 'Dog'
+    age: 4
+  }
+)
+```
+
+In `lib/slack_api_wrapper.rb` we will use this method to post data to Slack using the Slack URL and the API key we saved in our `.env` file.
+
+```ruby
+class SlackApiWrapper
+  BASE_URL = 'https://slack.com/api/'
+  API_KEY = ENV['SLACK_TOKEN']
+
+  def self.send_msg(message, channel)
+
+    response = HTTParty.post(
+      "#{BASE_URL}/chat.postMessage",
+      body:  {
+        token: API_KEY,
+        text: message,
+        channel: channel
+      },
+      headers: { 'Content-Type' => 'application/x-www-form-urlencoded' }
+    )
+
+    response_body = JSON.parse(response.body)
+
+    return response.success? && response_body["ok"]
```

Comment:
It might be worth spending a sentence to describe what's going on here.

I spent some time looking for some docs on what `.success?` does and actually didn't find.... basically any worthwhile documentation on HTTParty. Eugh.

This is the best I got in terms of "gives a decent explanation and example," but ugh, it's REALLY bad official documentation for HTTParty
https://stackoverflow.com/questions/7909596/how-can-i-handle-errors-with-httparty
https://www.rubydoc.info/github/jnunemaker/httparty/master/HTTParty/Response